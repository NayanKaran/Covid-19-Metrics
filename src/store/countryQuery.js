import { createSlice } from '@reduxjs/toolkit';

const slice = createSlice({
  name: 'countryQuery',
  initialState: {},
  reducers: {
    setQuery(query, action) {
      /* eslint-disable no-param-reassign */
      query[action.payload.continent] = action.payload.query;
    },
  },
});

export const { setQuery } = slice.actions;

export default slice.reducer;
