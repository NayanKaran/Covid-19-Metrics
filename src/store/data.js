import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import getCoronavirusData from '../modules/coronavirusAPI';

export const fetchCoronavirusDataFromAPI = createAsyncThunk(
  'data/fetchCoronavirusDataFromAPI',
  async () => {
    const response = await getCoronavirusData();
    return response;
  },
);

const slice = createSlice({
  name: 'data',
  initialState: null,
  extraReducers: (builder) => {
    builder.addCase(
      fetchCoronavirusDataFromAPI.fulfilled,
      (data, action) => action.payload,
    );
  },
});

export default slice.reducer;
