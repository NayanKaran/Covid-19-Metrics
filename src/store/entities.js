import { combineReducers } from '@reduxjs/toolkit';
import dataReducer from './data';
import queryReducer from './query';

export default combineReducers({
  data: dataReducer,
  query: queryReducer,
});
