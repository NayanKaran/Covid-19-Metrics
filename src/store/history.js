import { createSlice } from '@reduxjs/toolkit';

const slice = createSlice({
  name: 'history',
  initialState: '/',
  reducers: {
    setHistory(query, action) {
      /* eslint-disable no-param-reassign */
      return action.payload;
    },
  },
});

export const { setHistory } = slice.actions;

export default slice.reducer;
