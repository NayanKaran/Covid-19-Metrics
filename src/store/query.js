import { createSlice } from '@reduxjs/toolkit';

const slice = createSlice({
  name: 'query',
  initialState: '',
  reducers: {
    setQuery(query, action) {
      /* eslint-disable no-param-reassign */
      return action.payload;
    },
  },
});

export const { setQuery } = slice.actions;

export default slice.reducer;
