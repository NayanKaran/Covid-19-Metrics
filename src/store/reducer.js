import { combineReducers } from 'redux';
import entitiesReducer from './entities';
import historyReducer from './history';

export default combineReducers({
  entities: entitiesReducer,
  route: historyReducer,
});
