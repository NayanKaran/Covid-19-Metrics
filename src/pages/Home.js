/* eslint-disable global-require */
import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { fetchCoronavirusDataFromAPI } from '../store/data';
import { setQuery } from '../store/continentQuery';

let apiCalled = false;

const Home = () => {
  const data = useSelector((state) => state.entities.data);
  const query = useSelector((state) => state.entities.continentQuery);
  const dispatch = useDispatch();

  if (apiCalled === false && data === null) {
    apiCalled = true;
    dispatch(fetchCoronavirusDataFromAPI());
  }
  if (data) {
    const listItems = [];

    Object.keys(data.global).forEach((key) => {
      if (key !== 'all' && key.toLowerCase().includes(query.toLowerCase())) {
        listItems.push(
          <div
            key={key}
            className="continentListItem"
            style={{
              backgroundImage: `url('./images/continents/${key.toLowerCase().replace(' ', '_')}.svg')`,
            }}
          >
            <Link to={`/${key}`}>
              <i className="fa-regular fa-circle-right" style={{ color: 'white' }} />
            </Link>
            <p>{key}</p>
            <p>{`${data.global[key].all}`}</p>
          </div>,
        );
      }
    });

    const handleQuery = (event) => {
      if (query !== event.target.value) dispatch(setQuery(event.target.value));
    };

    return (
      <main>
        <div className="global">
          <img
            className="globalIcon"
            src="./images/global_map.svg"
            alt="globe icon"
            width={200}
          />
          <div className="globalInfo">
            <p className="globalTitle">Global</p>
            <p className="globalCases">{`${data.global.all} cases`}</p>
          </div>
        </div>
        <div className="searchBar">
          <input name="query" value={query} onChange={handleQuery} placeholder="Search continent" />
        </div>
        <div className="continentHeading">
          <h3>Continent breakdown</h3>
        </div>
        <ul className="continentList">{listItems}</ul>
      </main>
    );
  }
  return (
    <>
      <p>Loading...</p>
    </>
  );
};

export default Home;
