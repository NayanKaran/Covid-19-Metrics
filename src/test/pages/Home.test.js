import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Home from '../../pages/Home';
import store from '../../store/configureStore';

describe('Testing Home component', () => {
  it('Testing with snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Home />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});