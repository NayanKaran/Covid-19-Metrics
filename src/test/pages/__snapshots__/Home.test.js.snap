// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Home component Testing with snapshot 1`] = `
<p>
  Loading...
</p>
`;
